### ISSUE TITLE
Performance of converting to datatable is bad

### ISSUE THREAD

**Issue by AllNamesRTaken** [microsoft/ClearScript#254]
Hello, 
Great library!
But I make calls to APIs and build data in the form of a table that I return to the .NET host.
The format of the data is something like `{[key: string]: string | number | bool | null | undefined}[]`
This format is straight forward to convert into a DataTable but the performance is abysmal due to slow read of property data.
I guess it is due to the following code in the GetProperty call: 
`engine.MarshalToHost(engine.ScriptInvoke(() => target.GetProperty(index)), false);`
Is there a way to tell the engine to return somthing like Array<Dictionary<object>> or just not a DynamicObject?

If not, are there workarounds or plans to solve this?

**Comment by ClearScriptLib** [comment]
Hi @AllNamesRTaken,

Thanks for your kind words!

If we understand correctly, you're building an array of dictionaries in script code and passing it to a .NET method for processing. The .NET method uses a separate dynamic call to retrieve each dictionary entry. This results in poor performance due to a large number of hops across the host-script boundary.

If that's correct, our first suggestion would be as follows. In script code, once you've constructed your data, transform it into an array of [JSON](https://www.json.org/json-en.html) strings using [`JSON.stringify`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify). Pass the result to your .NET method. On the .NET side, use a library like [Newtonsoft.Json](https://www.nuget.org/packages/Newtonsoft.Json/) to deserialize each JSON string into a .NET dictionary (see [here](https://www.newtonsoft.com/json/help/html/DeserializeDictionary.htm)).

If performance is still not up to par – for example, if your array is so large that retrieving each JSON string separately is a problem – you should be able to convert _all_ the data into one JSON string and pass it to .NET in a single hop.

It's also possible that your data is so large that converting it all into a JSON string is impossible. In that case, you might have to get more creative. One possibility might be to use a [JavaScript typed array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Typed_arrays) as your data transfer medium. ClearScript offers fast copying to and from typed arrays.

Thoughts?

**Comment by AllNamesRTaken** [comment]
Thank you for responding so quickly.
I believe you hit the nail on it's head. The JSON.stringify + json.net convert is the solution i use as workaround. But for large tables it is still a costly solution when compared to the speed of V8, which is why i had hoped for a solution where i.e. leafs / object properties with basic data types could have their values automatically converted. That way the marshaling would only happen per row in a datatable. 

Is that of interest of a bad idea?

**Comment by ClearScriptLib** [comment]
Hi @AllNamesRTaken,

>That way the marshaling would only happen per row

Please clarify. Our understanding is that each dictionary in the source data ends up being a row in the final table. If that's correct, and if you're transferring each dictionary as a JSON string, you're already marshaling once per row.

>But for large tables it is still a costly solution when compared to the speed of V8

Since you're starting with a pure JavaScript data structure, marshaling is unavoidable.

As a general principle, ClearScript favors proxy marshaling over automatic conversion. The goal is to avoid data loss, but as you've noticed, accessing data across the host-script boundary is expensive.

In performance-sensitive scenarios such as yours, where that expense is unacceptable, the best solution is to alter your data access patterns to minimize hops across that boundary. By switching to JSON strings for the dictionaries, you've taken a big step in that direction (incidentally, we'd love to get an idea of the improvement it yielded over your original approach).

Beyond that, here are some ideas for further gains:

1. Transfer your entire array as a single JSON string. The performance gain is likely to depend on the array size.

1. Can you think of any way to optimize your data pre-transfer? For example, do the rows in your final table use a common schema? If so, it may be more efficient to transfer the data as a list of value arrays rather than dictionaries.

1. Use a format that's more efficient than JSON. Standard JavaScript only supports JSON, but there are libraries out there for things like [UBJSON](https://ubjson.org/) and [MessagePack](https://msgpack.org/index.html). If you serialized your data to a binary buffer ([Uint8Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array) et al), ClearScript could transfer it to a .NET array very efficiently. On the other hand, it's possible that V8's native JSON serializer is _faster_ than anything one could code in JavaScript, even if the resulting data is _larger_.

Finally, as you suggested, ClearScript could _natively_ implement some form of fast structured data transfer. That's something we've had on our backlog for a while, and we'll definitely look into it for the next release.

Please send any additional thoughts or findings!

Thanks!

**Comment by AllNamesRTaken** [comment]
Hi,
The data was about 2400 rows long and 19 columns wide. We returned it as a js array of objects, with each row  being one js object with 19 keys. Values were a mix of number and string types.

The original solution was to loop over the dynamic object representing the array and creating DataRows in a DataTable, one per dynamic representing the row object. Then per row looping over the keys and filling the cells with values. It was obvoius from VS perfomance metrics that the property access was the performance drain.
Total time on my machine debugging, though fluctuating, for this method was about 200- 300ms.

The current method is to JSON.serialize the whole array of objects in V8 and the do a Deserialize to Datatable with json.net.
Total time in debug mode for json.net was about 50ms, pretty deterministic.

For 2400×19 cells this feels slow.
Optimally would be iterating over the data as a pure array of dictionaries or something similar but i can understand the goal of not marshalling automatically to avoid errors from type conversion. But many usecases are not in the risk zone of these errors.

The middleground i was pondering was to automatically translate basic value properties such as string number,null and bool on objects. This would then reduce the need to marshall on every propery access when transforming into a datatable. If you wish to still allow for pure access to v8 types there could be a separate function on the clearscript object to access the pre translated data which would fallback to marshalling if nu value was found.

This would reduce the time in my example close to 20 times  for a more reasonable total of 10-15ms. The wet dream ofc would be a way to ask for automatic translation of arrays as well which would allow for very large data sets at low overhead.

Regards
Joel

**Comment by ClearScriptLib** [comment]
Hi Joel,

Thanks for providing that information! We'll run some experiments and get back to you. Hopefully your current JSON-based solution is good enough in the short term.

>The middleground i was pondering was to automatically translate basic value properties such as string number,null and bool on objects. This would then reduce the need to marshall on every propery access when transforming into a datatable.

It would appear that, by "translate", you mean transfer all primitive-valued properties at proxy construction, so that retrieving them no longer requires a round trip to the script engine. Is that correct? If so, it's an interesting idea; the problem is that access to those properties would no longer be "live", and the proxy could diverge from the original JavaScript object unless some coherency protocol were in place.

Quick question: We'd like to provide some kind of fast structured data transfer, but for practical reasons we're thinking of limiting it to JSON-compatible data. In your case, that would exclude `undefined` as a legal property value. Would that be a big problem for you?

Thanks again!

**Comment by AllNamesRTaken** [comment]
Yes you understood me right and yes i see the issue with stale access to
underlying data. I just believe the usecase to be common where you evaluate
to retrieve some object value to continue processing in .net. Perhaps a
parameter to the evaluate function or a separate call that do not provide
live access would be feasable?

Yes that would be enough for me with non-undefined basic types. My use case
is based around retrieving and processing larger data load over rest apis
which return json structures without the undefined type anyhow.

JSON serialization is doable for me in the short term but I would gladly
test any impovements that come at a later stage.

Many thanks for a positive duscussion. If I can be pf any assistance with
devekipment or testing please let me know.

Regards
Joel

Den fre 30 apr. 2021 22:36ClearScript Library ***@***.***>
skrev:
